warehouse/integrations/github/utils.py

```diff
@@ -0,0 +1,355 @@
+                data=public_key.encode("utf-8"), backend=default_backend()
+            )
+            loaded_public_key.verify(
+                signature=base64.b64decode(signature),
+                data=payload.encode("utf-8"),
+                # This validates the ECDSA and SHA256 part
+                signature_algorithm=ECDSA(algorithm=SHA256()),
+            )
+        except InvalidSignature as exc:
+            raise InvalidTokenLeakRequest(
+                "Invalid signature", "invalid_signature"
+            ) from exc
+        except Exception as exc:
+            # Maybe the key is not a valid ECDSA key, maybe the data is not properly
+            # padded, etc. So many things can go wrong...
+            raise InvalidTokenLeakRequest(
+                "Invalid cryptographic values", "invalid_crypto"
+            ) from exc
+
+
+def _analyze_disclosure(request, disclosure_record, origin):
+
+    metrics = request.find_service(IMetricsService, context=None)
+
+    metrics.increment(f"warehouse.token_leak.{origin}.recieved")
+
+    try:
+        disclosure = TokenLeakDisclosureRequest.from_api_record(
+            record=disclosure_record
+        )
+    except InvalidTokenLeakRequest as exc:
+        metrics.increment(f"warehouse.token_leak.{origin}.error.{exc.reason}")
+        return
+
+    macaroon_service = request.find_service(IMacaroonService, context=None)
+    try:
+        database_macaroon = macaroon_service.find_from_raw(
+            raw_macaroon=disclosure.token
+        )
+    except InvalidMacaroon:
+        metrics.increment(f"warehouse.token_leak.{origin}.error.invalid")
+        return
+
+    metrics.increment(f"warehouse.token_leak.{origin}.valid")
+
+    macaroon_service.delete_macaroon(macaroon_id=str(database_macaroon.id))
+
+    send_token_compromised_email_leak(
+        request,
+        database_macaroon.user,
+        public_url=disclosure.public_url,
+        origin=origin,
+    )
+    metrics.increment(f"warehouse.token_leak.{origin}.processed")
+
+
+def analyze_disclosure(request, disclosure_record, origin):
+    try:
+        _analyze_disclosure(
+            request=request,
+            disclosure_record=disclosure_record,
+            origin=origin,
+        )
+    except Exception:
+        metrics = request.find_service(IMetricsService, context=None)
+        metrics.increment(f"warehouse.token_leak.{origin}.error.unknown")
+        raise
+
+
+def analyze_disclosures(disclosure_records, origin, metrics):
+    from warehouse.integrations.github import tasks
+
+    if not isinstance(disclosure_records, list):
+        metrics.increment(f"warehouse.token_leak.{origin}.error.format")
+        raise InvalidTokenLeakRequest("Invalid format: payload is not a list", "format")
+
+    for disclosure_record in disclosure_records:
+        tasks.analyze_disclosure_task.delay(
+            disclosure_record=disclosure_record, origin=origin
+        )
```

Comment:
task needs to be wrapped in `task()` before calling delay. But I've yet to figure out where `task()` should come from

Reply:
Done